homeassistant/components/light/group.py

```diff
@@ -0,0 +1,262 @@
+    SUPPORT_WHITE_VALUE, PLATFORM_SCHEMA)
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'group'
+
+DEFAULT_NAME = 'Group Light'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required(CONF_ENTITIES): cv.entity_ids,
+})
+
+SUPPORT_GROUP_LIGHT = (SUPPORT_BRIGHTNESS | SUPPORT_COLOR_TEMP | SUPPORT_EFFECT
+                       | SUPPORT_FLASH | SUPPORT_RGB_COLOR | SUPPORT_TRANSITION
+                       | SUPPORT_XY_COLOR | SUPPORT_WHITE_VALUE)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass: HomeAssistantType,
+                         config: ConfigType,
+                         async_add_devices,
+                         discovery_info=None) -> None:
+    """Initialize light.group platform."""
+    async_add_devices(
+        [GroupLight(hass, config.get(CONF_NAME), config[CONF_ENTITIES])])
+
+
+class GroupLight(light.Light):
+    """Representation of a group light."""
+
+    def __init__(self, hass: HomeAssistantType, name: str,
+                 entity_ids: List[str]) -> None:
+        """Initialize a group light."""
+        self.hass = hass  # type: HomeAssistantType
+        self._name = name  # type: str
+        self._entity_ids = entity_ids  # type: List[str]
+        self._state = STATE_UNAVAILABLE  # type: str
+        self._brightness = None  # type: Optional[int]
+        self._xy_color = None  # type: Optional[Tuple[float, float]]
+        self._rgb_color = None  # type: Optional[Tuple[int, int, int]]
+        self._color_temp = None  # type: Optional[int]
+        self._min_mireds = 154  # type: Optional[int]
+        self._max_mireds = 500  # type: Optional[int]
+        self._white_value = None  # type: Optional[int]
+        self._effect_list = None  # type: Optional[List[str]]
+        self._effect = None  # type: Optional[str]
+        self._supported_features = 0  # type: int
+
+    @asyncio.coroutine
+    def async_added_to_hass(self) -> None:
+        """Register callbacks"""
+
+        @callback
+        def async_state_changed_listener(self, entity_id: str,
+                                         old_state: State, new_state: State):
+            """Handle child updates."""
+            self.async_schedule_update_ha_state(True)
+
+        @callback
+        def async_homeassistant_start(event):
+            """Track child state changes on startup."""
+            async_track_state_change(self.hass, self._entity_ids,
+                                     async_state_changed_listener)
+
+            self.async_schedule_update_ha_state(True)
+
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START,
+                                        async_homeassistant_start)
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def state(self) -> str:
```

Comment:
Just embed it inside `is_on`. Function calls add overhead. Also, we can just store a boolean since we don't ever use the value of self.state

Reply:
This is kind of because of my own way of using Home Assistant, but: Don't we want to also support the light reporting an `STATE_UNAVAILABLE` state when all child lights are `unavailable`?

> Just embed it inside is_on. Function calls add overhead. 

Do you mean writing this? (usage of `_`)

```python
@property
def is_on(self) -> bool:
    """Return True if entity is on."""
    return self._state == STATE_ON
```

Reply:
`return self._is_on`

Reply:
(sorry if this is outdated: we'd still like to report `unavailable`/`unknown` state right? Then AFAIK we can't *just* implement the property `is_on`. We need `Light#state` for that.)

Reply:
You are **never** allowed to override `Light#state` or any `state` property of any abstract base class for that matter. You should always implement properties and what gets exposed as state is up to the ABC to decide. You implement `is_on` and `available`, the ABC decides how that will be represented.

Reply:
👍